apis/python/src/tiledbsoma/_dataframe.py

```diff
@@ -574,41 +576,18 @@ def _set_reader_coord_by_py_seq_or_np_array(
                     f"only 1D numpy arrays may be used to index; got {coord.ndim}"
                 )
 
-        # See libtiledbsoma.cc for more context on why we need the
-        # explicit type-check here.
-
-        if dim.dtype == np.int64:
-            sr.set_dim_points_int64(dim.name, coord)
-        elif dim.dtype == np.int32:
-            sr.set_dim_points_int32(dim.name, coord)
-        elif dim.dtype == np.int16:
-            sr.set_dim_points_int16(dim.name, coord)
-        elif dim.dtype == np.int8:
-            sr.set_dim_points_int8(dim.name, coord)
-
-        elif dim.dtype == np.uint64:
-            sr.set_dim_points_uint64(dim.name, coord)
-        elif dim.dtype == np.uint32:
-            sr.set_dim_points_uint32(dim.name, coord)
-        elif dim.dtype == np.uint16:
-            sr.set_dim_points_uint16(dim.name, coord)
-        elif dim.dtype == np.uint8:
-            sr.set_dim_points_uint8(dim.name, coord)
-
-        elif dim.dtype == np.float64:
-            sr.set_dim_points_float64(dim.name, coord)
-        elif dim.dtype == np.float32:
-            sr.set_dim_points_float32(dim.name, coord)
-
-        elif dim.dtype == "str" or dim.dtype == "bytes":
-            sr.set_dim_points_string_or_bytes(dim.name, coord)
+        try:
+            set_dim_points = getattr(sr, f"set_dim_points_{dim.type}")
+        except AttributeError:
+            # We have to handle this type specially below
+            pass
+        else:
+            set_dim_points(dim.name, coord)
+            return True
 
-        elif (
-            dim.dtype == "datetime64[s]"
-            or dim.dtype == "datetime64[ms]"
-            or dim.dtype == "datetime64[us]"
-            or dim.dtype == "datetime64[ns]"
-        ):
+        if _util.pa_types_is_string_or_bytes(dim.type):
+            sr.set_dim_points_string_or_bytes(dim.name, coord)
```

Comment:
would recommend the structure

```python
if condition:
  do_thing()
  return True
if other_condition():
  ...
  return True
# fell through everything
raise ValueError(...)
```

so that the reader doesn’t have to go past all the if-elses to find the return (and the `raise` doesn’t need to live in its own `else` either).